Fix bug in drawPixel asm

Pharap opened the pull request:
The processor was locking up because y wasn't writable.
The solution was to make y a read-write (`+`) output and do the cast to `uint8_t` in assembly.

MLXXXp commented:
Is there an explanation for why it worked with the _Arduboy2_ class but not the _Arduboy2Base_ class?

Pharap commented:
I couldn't find a reason. I'm assuming that in `Arduboy2Base` the compiler was picking a register that didn't work, but in `Arduboy2` it was picking a viable register. 

(Sidenote: I did a test a while back that seemed to indicate that GCC decides to duplicate inherited functions rather than passing a pointer of the child class to the parent's function, which may be related to difference in struct sizes, packing or alignment.)

Ultimately though, the mystery isn't why Arduboy2Base didn't work, it's why Arduboy2 did.
Gcc explicitly states in its [inline assembly guide](https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html#InputOperands):

> **Warning:** Do not modify the contents of input-only operands (except for inputs tied to outputs). The compiler assumes that on exit from the asm statement these operands contain the same values as they had before executing the statement. It is not possible to use clobbers to inform the compiler that the values in these inputs are changing. One common work-around is to tie the changing input variable to an output variable that never gets used. Note, however, that if the code that follows the asm statement makes no use of any of the output operands, the GCC optimizers may discard the asm statement as unneeded 

Which is what lead me to try the solution that worked.

MLXXXp commented:
@Pharap, Your fix makes sense but is the
"andi %[y], 0xFF\n"
instruction necessary to make it work? (It doesn't look like it's required, to me.) Doing an 8 bit _and_ with 0xFF is essentially an NOP, so is wasting code size and speed.

Also, please modify your commit message to follow [50/72 guidelines](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html):

- Change _Fixed_ to _Fix_
- Wrap the long line to 72 characters or less

MLXXXp commented:
I wonder if it's more proper to use _A_ to specify that it's the low byte of _y_ that's being used (that _y_ is being cast to _uint8_t_)?

```
"mul %[width_offset],%A[y]\n"
...
"andi %A[y], 0x07\n"
"add r30, %A[y]\n"
```

It doesn't appear to affect the resulting output but could make it more understandable.

Pharap commented:
Ah, I thought it was a 2-byte operation (and thus equivalent to a cast from `uint16_t` to `uint8_t`).
I double-checked the documentation and it seems that `mul` is an 8-bit operation as well, which means that the original cast was also unnecessary.
I'll remove that line when I've fixed the current commit title.

I had a look but I couldn't find any guidelines on your readme or anywhere else, you should probably mention that somewhere significant in case someone else tries to provide a bug fix.

I can't seem to find which of the lines I added is breaking the 72-character limit.
I can see one line, but it's a pre-existing line, not one I added.

I can ammend it but it might take some time as I'll have to install git and find out how to perform the ammendment.

I didn't add the A because I couldn't find any documentation explaining what the `A`s were actually for (I did in fact try removing them, but the result didn't appear to be any different, so I assumed they were a notation quirk and left them alone). Technically speaking (after seeing that `andi` and `mul` both work on byte-sized operands) `y` is never actually cast to `uint8_t`, it's just a matter of its high byte going unused.

MLXXXp commented:
@Pharap ,
> I can't seem to find which of the lines I added is breaking the 72-character limit.

The 50/72 guideline is for the _commit message_, not the actual code. It's the line (longer than 72 characters):
`The solution was to make y a read-write (`+`) output and do the cast to uint8_t in assembly.`

> I'll have to install git and find out how to perform the ammendment.

Once you've installed git:

- git clone https://github.com/Pharap/Arduboy2
- git commit --amend
- git push --force origin master

> I didn't add the A because I couldn't find any documentation explaining what the As were actually for

_A_ means the least significant byte, _B_ means the 2nd least significant byte, _C_ the third, etc.
http://www.nongnu.org/avr-libc/user-manual/inline_asm.html

I'm now wondering if we have the same potential problem with *bitshift_left* since it's declared as an input operand but we're writing to it? It's the _Z_ register, which is also R30 and R31.

Pharap commented:
> > The 50/72 guideline is for the commit message, not the actual code. It's the line (longer than 72 characters):
> The solution was to make y a read-write (+) output and do the cast to uint8_t in assembly.

That makes much more sense. Consider that done.

> Once you've installed git:
> 
> -     git clone https://github.com/Pharap/Arduboy2
> -     git commit --amend
> -     git push --force origin master

I had already cloned successfully, it was `amend` that I was stuck on because the default editor was some weird command line thing that had no obvious way of accepting something as 'finished'. Luckily I discovered that PortableGit comes with a Gui for Windows, so I used that instead.

> A means the least significant byte, B means the 2nd least significant byte, C the third, etc.
> http://www.nongnu.org/avr-libc/user-manual/inline_asm.html

Ah, good to know. I didn't read the lower parts because they just looked like examples and wasn't expecting them to be outlining different features.

> I'm now wondering if we have the same potential problem with bitshift_left since it's declared as an input operand but we're writing to it? It's the Z register, which is also R30 and R31.

It's possible. I'm thinking that maybe it hasn't been having issues because it's a special purpose register (like `x` and `y`) whereas `d` could resolve to any of the upper registers including the 5 non-special pairs.
Regardless of why it hasn't been an issue so far I think it would probably be best to change it to an output purely because of GCC's warning against overwriting inputs. I shall make that ammendment now, along with marking `y` as `A`.

MLXXXp commented:
> Regardless of why it hasn't been an issue so far I think it would probably be best to change it to an output purely because of GCC's warning against overwriting inputs.

I haven't been able find a way to do that without getting an error. I'm beginning to think that it's safe to leave it as is because Z is a special register; it's the only one that the `lpm` instruction can use.

Pharap commented:
> I haven't been able find a way to do that without getting an error. I'm beginning to think that it's safe to leave it as is because Z is a special register; it's the only one that the lpm instruction can use.

`bitshift_left` is `const` and therefore cannot be assigned to a writable register.
To make `z` writable set `bitshift_left` as an input to a different register and `movw` it into `z`.
However this does mean that `z` could be set as a clobber rather than an output.
e.g.
```
    : [row_offset] "=x&" (row_offset), // upper register (ANDI)
      [bit] "=r" (bit),
      [y] "+d" (y) // upper register (ANDI), must be writable
    : [width_offset] "r" ((uint8_t)(WIDTH/8)),
      [x] "r" ((uint8_t)x),
      [bitshift_left] "r" (bitshift_left)
    : "r30", "r31"
```

It means an extra instruction, but it works and it's future proof.
I won't commit it yet in case you decide it's better to simply carry on as before and hope the specialness protects `z`.

MLXXXp commented:
I think we're safe leaving *bitshift_left* as an input even though we're changing it. This is because:

- the ["FAQ: What registers are used by the C compiler?"](http://www.nongnu.org/avr-libc/user-manual/FAQ.html#faq_reg_usage) states:
> - **Call-used registers (r18-r27, r30-r31):**
May be allocated by gcc for local data. You _may_ use them freely in assembler subroutines. Calling C subroutines can clobber any of them - the caller is responsible for saving and restoring.
 
Even though the above FAQ is talking about entire functions written in assembler, not inline assembler within a C function, it still indicates that _Z (R30 and R31)_ can be changed within the function without regards to other functions.

- The *bitshift_left* array is only accessed within the assembly code.  As long as we make sure we don't use *bitshift_left* again, any other future use of _Z_ within the function will have to load it first with a new value, since this new value won't be *bitshift_left*, which it knows was declared to be in _Z_.

Pharap commented:
I'm a bit concerned about the "May be allocated by gcc for local data." part, that sounds to me like GCC is allowed to write over it which indicates it's prone to the same issue that affected `y`, but I'll leave that as-is.

I don't think reusing `bitshift_left` is going to be an issue if you do it in a separate `asm` block because the input specifiers translate to a `mov` or `movw` anyway, so if you're declaring `"z" (bitshift_left)` the previous value of `z` is overwritten. Even if you're not using `z` as an input you'd have to overwrite it in any case, no matter what code came before it - as that FAQ said, any C function could clobber it.

Is there anything else that needs to be added or is the current code acceptable?

MLXXXp commented:
> I don't think reusing `bitshift_left` is going to be an issue if you do it in a separate asm block

I did a test where I added a separate asm block at the end of the function with, `"z" (bitshift_left)` declared again as an input operator. The compiler blindly assumed that _Z_ still contained the value *bitshift_left*. It didn't reload it. So, it **is** an issue.

I may ask around to see if there's a proper way of doing this without needing to use a working register in addition to _Z_.

> Is there anything else that needs to be added or is the current code acceptable?

I want to test your changes a bit myself. If I don't find any problems, I think the code is acceptable.

It would be nice if you could squash the 3 commits into one using _git rebase_. If you do, could you also add a space after the comma between the _mul_ instruction operands, for consistency? (I know this wasn't your doing.)

Pharap commented:
> I may ask around to see if there's a proper way of doing this without needing to use a working register in addition to Z.

I suspect the proper way is to use another register.
That or maybe  use a [Local Register Variable](https://gcc.gnu.org/onlinedocs/gcc/Local-Register-Variables.html#Local-Register-Variables) as a hack to reinitialise the `z` register.
E.g. `register uint8_t * tempZ asm ("z") = bitshift_left;`
I don't know if that's guaranteed though.

> It would be nice if you could squash the 3 commits into one using git rebase. If you do, could you also add a space after the comma between the mul instruction operands, for consistency? (I know this wasn't your doing.)

It was a bit fiddly because of a merge conflict (I have discovered I really don't like vim or vimdiff) but I managed it.

MLXXXp commented:
@Pharap, I don't know what you did but this PR is now totally messed up.

Pharap commented:
Oh, I see what's happened. The rebasing went wrong

MLXXXp commented:
Click on _Files changed_ at the top of this PR to see a diff of your proposed changes.

Pharap committed:
Fix bug in drawPixel asm

Pharap commented:
The former mistake was due to the rebasing behaving differently to how I expected. I was expecting it to merge the files together, not entirely delete them.

I ended up accidentally closing this PR by doing another rebase, but I effectively undid everything I'd done and then recommitted everything at once. (Had to do it twice because I accidentally left in an extra line that I'd commented out while testing.)

It definitely 100% works this time.
I've compiled it, tested it, and triple checked that it's the same on github as it is in the local repo and my working copy.

joshgoebel commented:
Ok, I get the core problem here, but why is %A necessary?

MLXXXp commented:
I don't know if using _%A[y]_, instead of just _%[y]_, is strictly necessary in this case but _y_ is a 16 bit value and from a C++ point of view we want to cast it as a *uint8_t*. Using _%A[y]_ makes it clear that it's the lower 8 bits of _y_ that we want to use.
